File: packages/python/src/tesote_sdk/transport.py
"""Single HTTP client for every resource module across v1/v2.

Owns: bearer injection, retries with exponential backoff + jitter, rate-limit
header capture, idempotency-key auto-generation, request-id propagation into
errors, opt-in TTL LRU cache, bearer redaction in logs and error summaries.

Stdlib only -- ``urllib.request`` + ``json``. Zero runtime dependencies.
"""

from __future__ import annotations

import json
import socket
import ssl
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
import uuid
from typing import Any, Callable, Dict, Mapping, Optional

from ._internals import (
    CacheBackend,
    InMemoryLRUCache,
    JsonType,
    LoggerCallback,
    RateLimitSnapshot,
    Response,
    RetryPolicy,
    redact_bearer,
)
from ._version import __version__
from .errors import (
    ApiError,
    NetworkError,
    TesoteError,
    TimeoutError,
    TlsError,
    TransportError,
    classify_api_error,
)

DEFAULT_BASE_URL = "https://equipo.tesote.com/api"
DEFAULT_CONNECT_TIMEOUT = 5.0
DEFAULT_READ_TIMEOUT = 30.0

_MUTATING_METHODS = frozenset({"POST", "PUT", "PATCH", "DELETE"})
_RETRY_STATUS = frozenset({429, 502, 503, 504})


def _sleep(seconds: float) -> None:  # why: indirection makes tests deterministic
    if seconds > 0:
        time.sleep(seconds)


# why: kept for back-compat with tests that import this private helper directly
_redact_bearer = redact_bearer


class Transport:
    """Single HTTP client used by every versioned resource module."""

    def __init__(
        self,
        api_key: str,
        *,
        base_url: str = DEFAULT_BASE_URL,
        connect_timeout: float = DEFAULT_CONNECT_TIMEOUT,
        read_timeout: float = DEFAULT_READ_TIMEOUT,
        retry_policy: Optional[RetryPolicy] = None,
        cache_backend: Optional[CacheBackend] = None,
        user_agent: Optional[str] = None,
        logger: Optional[LoggerCallback] = None,
        opener: Optional[Callable[..., Any]] = None,
    ) -> None:
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._connect_timeout = connect_timeout
        self._read_timeout = read_timeout
        self._retry_policy = retry_policy or RetryPolicy()
        self._cache_backend = cache_backend
        self._user_agent = user_agent or self._default_user_agent()
        self._logger = logger
        # why: caller-injected opener allows tests to swap urlopen without monkeypatching
        self._opener = opener or urllib.request.urlopen
        self.last_rate_limit: Optional[RateLimitSnapshot] = None

    @staticmethod
    def _default_user_agent() -> str:
        v = sys.version_info
        return f"tesote-sdk-py/{__version__} (python/{v.major}.{v.minor}.{v.micro})"

    @property
    def base_url(self) -> str:
        return self._base_url

    def request(
        self,
        method: str,
        path: str,
        *,
        query: Optional[Mapping[str, Any]] = None,
        body: Optional[Mapping[str, Any]] = None,
        idempotency_key: Optional[str] = None,
        cache_ttl: Optional[float] = None,
        extra_headers: Optional[Mapping[str, str]] = None,
    ) -> Response:
        """Execute one logical request, including retries.

        Returns a parsed :class:`Response`. On a non-2xx that we do not retry
        (or retries exhausted), raises the typed :class:`ApiError` subclass.
        """
        method_upper = method.upper()
        full_url = self._build_url(path, query)
        headers = self._build_headers(method_upper, idempotency_key, extra_headers)

        cache_key = self._cache_key(method_upper, full_url) if cache_ttl else None
        if cache_key and self._cache_backend is not None:
            cached = self._cache_backend.get(cache_key)
            if cached is not None:
                _, response = cached
                return response

        encoded_body = json.dumps(body).encode("utf-8") if body is not None else None
        request_summary = self._summarize_request(method_upper, path, query, body)

        attempt = 0
        last_error: Optional[TesoteError] = None
        while attempt < self._retry_policy.max_attempts:
            attempt += 1
            try:
                response = self._perform(method_upper, full_url, headers, encoded_body)
            except TesoteError as exc:
                exc.attempts = attempt
                last_error = exc
                if not self._should_retry_error(exc, method_upper, idempotency_key):
                    raise
                _sleep(self._retry_policy.compute_delay(attempt))
                continue

            self._capture_rate_limit(response.headers)
            if 200 <= response.status < 300:
                if cache_key and self._cache_backend is not None and cache_ttl:
                    expires_at = time.monotonic() + cache_ttl
                    self._cache_backend.set(cache_key, (expires_at, response))
                return response

            api_error = self._build_api_error(
                response, request_summary=request_summary, attempts=attempt
            )
            last_error = api_error
            if response.status in _RETRY_STATUS and attempt < self._retry_policy.max_attempts:
                retry_after = self._parse_retry_after(response.headers)
                _sleep(self._retry_policy.compute_delay(attempt, retry_after))
                continue
            raise api_error

        # why: loop exited via `continue` only; surface the last typed error
        assert last_error is not None
        raise last_error

    # ------------------------------------------------------------------
    # internals
    # ------------------------------------------------------------------

    def _build_url(self, path: str, query: Optional[Mapping[str, Any]]) -> str:
        if not path.startswith("/"):
            path = "/" + path
        url = self._base_url + path
        if query:
            filtered = {k: v for k, v in query.items() if v is not None}
            if filtered:
                url = url + "?" + urllib.parse.urlencode(filtered, doseq=True)
        return url

    def _build_headers(
        self,
        method: str,
        idempotency_key: Optional[str],
        extra: Optional[Mapping[str, str]],
    ) -> Dict[str, str]:
        headers: Dict[str, str] = {
            "Authorization": f"Bearer {self._api_key}",
            "Accept": "application/json",
            "User-Agent": self._user_agent,
        }
        if method in _MUTATING_METHODS:
            headers["Content-Type"] = "application/json"
            key = idempotency_key or str(uuid.uuid4())
            headers["Idempotency-Key"] = key
        if extra:
            headers.update(extra)
        return headers

    @staticmethod
    def _cache_key(method: str, full_url: str) -> str:
        return f"{method} {full_url}"

    def _summarize_request(
        self,
        method: str,
        path: str,
        query: Optional[Mapping[str, Any]],
        body: Optional[Mapping[str, Any]],
    ) -> Dict[str, Any]:
        body_shape: Optional[Dict[str, Any]] = None
        if body is not None:
            body_shape = {"keys": sorted(body.keys()), "size": len(json.dumps(body))}
        return {
            "method": method,
            "path": path,
            "query": dict(query) if query else None,
            "body_shape": body_shape,
            # why: explicit redaction so a logger that captures the summary can't leak the key
            "authorization": redact_bearer(self._api_key),
        }

    def _perform(
        self,
        method: str,
        url: str,
        headers: Mapping[str, str],
        body: Optional[bytes],
    ) -> Response:
        req = urllib.request.Request(url=url, data=body, method=method)
        for k, v in headers.items():
            req.add_header(k, v)

        log_payload: Dict[str, Any] = {
            "phase": "request",
            "method": method,
            "url": url,
            "headers": {**headers, "Authorization": redact_bearer(self._api_key)},
        }
        self._log(log_payload)

        # why: urlopen `timeout` covers connect+read; the SDK exposes both for
        # symmetry with other languages -- we apply the larger of the two.
        timeout = max(self._connect_timeout, self._read_timeout)
        try:
            raw = self._opener(req, timeout=timeout)
        except urllib.error.HTTPError as exc:
            return self._read_http_error(exc)
        except urllib.error.URLError as exc:
            raise self._classify_url_error(exc) from exc
        except socket.timeout as exc:
            raise TimeoutError(
                f"Request timed out after {timeout}s",
            ) from exc
        except (ConnectionError, OSError) as exc:
            raise NetworkError(f"Network error: {exc}") from exc

        return self._read_http_response(raw)

    def _read_http_response(self, raw: Any) -> Response:
        status = int(getattr(raw, "status", None) or raw.getcode())
        headers = self._headers_to_dict(raw.headers)
        body_bytes: bytes = raw.read()
        try:
            raw.close()
        except Exception:  # noqa: BLE001 -- best-effort cleanup
            pass
        body_text = body_bytes.decode("utf-8", errors="replace")
        parsed = self._parse_json(body_text)
        request_id = headers.get("X-Request-Id") or headers.get("x-request-id")
        self._log(
            {
                "phase": "response",
                "status": status,
                "request_id": request_id,
                "headers": headers,
            }
        )
        return Response(status, headers, body_text, parsed, request_id)

    def _read_http_error(self, exc: urllib.error.HTTPError) -> Response:
        status = exc.code
        headers = self._headers_to_dict(exc.headers) if exc.headers else {}
        try:
            body_bytes = exc.read()
        except Exception:  # noqa: BLE001
            body_bytes = b""
        body_text = body_bytes.decode("utf-8", errors="replace")
        parsed = self._parse_json(body_text)
        request_id = headers.get("X-Request-Id") or headers.get("x-request-id")
        self._log(
            {
                "phase": "response",
                "status": status,
                "request_id": request_id,
                "headers": headers,
            }
        )
        return Response(status, headers, body_text, parsed, request_id)

    @staticmethod
    def _headers_to_dict(headers: Any) -> Dict[str, str]:
        if hasattr(headers, "items"):
            return {str(k): str(v) for k, v in headers.items()}
        return {}

    @staticmethod
    def _parse_json(text: str) -> JsonType:
        if not text:
            return None
        try:
            return json.loads(text)  # type: ignore[no-any-return]
        except json.JSONDecodeError:
            return None

    def _classify_url_error(self, exc: urllib.error.URLError) -> TransportError:
        reason = exc.reason
        if isinstance(reason, ssl.SSLError):
            return TlsError(f"TLS error: {reason}")
        if isinstance(reason, socket.timeout):
            return TimeoutError(f"Connection timed out: {reason}")
        return NetworkError(f"Network error: {reason}")

    def _capture_rate_limit(self, headers: Mapping[str, str]) -> None:
        def _int(name: str) -> Optional[int]:
            value = headers.get(name) or headers.get(name.lower())
            if value is None:
                return None
            try:
                return int(value)
            except ValueError:
                return None

        self.last_rate_limit = RateLimitSnapshot(
            limit=_int("X-RateLimit-Limit"),
            remaining=_int("X-RateLimit-Remaining"),
            reset=_int("X-RateLimit-Reset"),
        )

    @staticmethod
    def _parse_retry_after(headers: Mapping[str, str]) -> Optional[float]:
        value = headers.get("Retry-After") or headers.get("retry-after")
        if value is None:
            return None
        try:
            return float(value)
        except ValueError:
            return None

    @staticmethod
    def _should_retry_error(
        exc: TesoteError, method: str, idempotency_key: Optional[str]
    ) -> bool:
        if isinstance(exc, TimeoutError):
            # why: read-timeouts on non-idempotent calls may have succeeded server-side
            if method in _MUTATING_METHODS and idempotency_key is None:
                return False
            return True
        if isinstance(exc, NetworkError):
            return True
        return False

    def _build_api_error(
        self,
        response: Response,
        *,
        request_summary: Mapping[str, Any],
        attempts: int,
    ) -> ApiError:
        envelope: Dict[str, Any] = {}
        if isinstance(response.json, dict):
            envelope = response.json
        error_code = envelope.get("error_code")
        error_id = envelope.get("error_id")
        message = envelope.get("error") or f"HTTP {response.status}"
        retry_after = self._parse_retry_after(response.headers)
        if retry_after is None and isinstance(envelope.get("retry_after"), (int, float)):
            retry_after = float(envelope["retry_after"])

        cls = classify_api_error(error_code, response.status)
        return cls(
            message,
            error_code=error_code,
            http_status=response.status,
            request_id=response.request_id,
            error_id=error_id,
            retry_after=int(retry_after) if retry_after is not None else None,
            response_body=response.body,
            request_summary=request_summary,
            attempts=attempts,
        )

    def _log(self, payload: Mapping[str, Any]) -> None:
        if self._logger is None:
            return
        try:
            self._logger(payload)
        except Exception:  # noqa: BLE001 -- a misbehaving logger must not break a request
            pass


__all__ = [
    "Transport",
    "Response",
    "RetryPolicy",
    "RateLimitSnapshot",
    "CacheBackend",
    "InMemoryLRUCache",
    "LoggerCallback",
    "DEFAULT_BASE_URL",
    "DEFAULT_CONNECT_TIMEOUT",
    "DEFAULT_READ_TIMEOUT",
]

File: packages/ruby/lib/tesote_sdk/transport.rb
require 'json'
require 'net/http'
require 'openssl'
require 'securerandom'
require 'uri'

require_relative 'errors'
require_relative 'version'

module TesoteSdk
  RateLimitInfo = Struct.new(:limit, :remaining, :reset, keyword_init: true)

  # Minimal LRU + TTL cache. Duck-typed: any object responding to #read(key)
  # and #write(key, value, ttl:) can be passed via :cache_backend.
  class CacheBackend
    Entry = Struct.new(:value, :expires_at)

    def initialize(max_size: 256)
      @max_size = max_size
      @data = {}
      @mutex = Mutex.new
    end

    def read(key)
      @mutex.synchronize do
        entry = @data.delete(key)
        return nil if entry.nil?
        if entry.expires_at < monotonic_now
          return nil
        end

        @data[key] = entry
        entry.value
      end
    end

    def write(key, value, ttl:)
      @mutex.synchronize do
        @data.delete(key)
        @data[key] = Entry.new(value, monotonic_now + ttl)
        @data.shift while @data.size > @max_size
        value
      end
    end

    def clear
      @mutex.synchronize { @data.clear }
    end

    private

    def monotonic_now
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end
  end

  class Transport
    DEFAULT_BASE_URL = 'https://equipo.tesote.com/api'.freeze
    DEFAULT_OPEN_TIMEOUT = 5
    DEFAULT_READ_TIMEOUT = 30
    DEFAULT_MAX_ATTEMPTS = 3
    DEFAULT_BASE_DELAY = 0.250
    DEFAULT_MAX_DELAY = 8.0
    MUTATING_METHODS = %w[POST PUT PATCH DELETE].freeze
    RETRYABLE_STATUSES = [429, 502, 503, 504].freeze

    attr_reader :api_key, :base_url, :version_segment, :user_agent,
                :open_timeout, :read_timeout, :max_attempts, :base_delay,
                :max_delay, :logger, :cache_backend

    attr_accessor :last_rate_limit, :last_request_id

    def initialize(api_key:,
                   version_segment:,
                   base_url: DEFAULT_BASE_URL,
                   user_agent: nil,
                   open_timeout: DEFAULT_OPEN_TIMEOUT,
                   read_timeout: DEFAULT_READ_TIMEOUT,
                   max_attempts: DEFAULT_MAX_ATTEMPTS,
                   base_delay: DEFAULT_BASE_DELAY,
                   max_delay: DEFAULT_MAX_DELAY,
                   logger: nil,
                   cache_backend: nil,
                   sleeper: nil,
                   randomizer: nil)
      raise ConfigError, 'api_key is required' if api_key.nil? || api_key.to_s.empty?
      raise ConfigError, 'version_segment is required' if version_segment.nil? || version_segment.to_s.empty?

      @api_key = api_key.to_s
      @version_segment = version_segment.to_s
      @base_url = base_url.to_s.sub(%r{/+\z}, '')
      @user_agent = user_agent || default_user_agent
      @open_timeout = open_timeout
      @read_timeout = read_timeout
      @max_attempts = max_attempts
      @base_delay = base_delay
      @max_delay = max_delay
      @logger = logger
      @cache_backend = cache_backend
      @sleeper = sleeper || ->(seconds) { sleep(seconds) }
      @randomizer = randomizer || ->(max) { Kernel.rand(max) }
      @last_rate_limit = nil
      @last_request_id = nil
    end

    # opts:
    #   :idempotency_key      → forwarded as Idempotency-Key (auto-gen for mutations)
    #   :cache                → false to bypass; { ttl: int } to enable TTL cache
    #   :extra_headers        → hash of additional headers
    def request(method, path, query: nil, body: nil, opts: {})
      method_upper = method.to_s.upcase
      uri = build_uri(path, query)
      execute_request(method_upper, uri, body, opts)
    end

    # why: GET /status and GET /whoami live at the API root, not under
    # /v1 or /v2 — bypass the version_segment but reuse all cross-cutting.
    def request_unversioned(method, path, query: nil, body: nil, opts: {})
      method_upper = method.to_s.upcase
      uri = build_unversioned_uri(path, query)
      execute_request(method_upper, uri, body, opts)
    end

    # Returns a RawResponse with body string + headers — used for file-download
    # endpoints (CSV/JSON export) where the SDK should not parse the body.
    RawResponse = Struct.new(:status, :body, :content_type, :content_disposition, :request_id, keyword_init: true)

    def request_raw(method, path, query: nil, body: nil, opts: {})
      method_upper = method.to_s.upcase
      uri = build_uri(path, query)
      request_summary = build_request_summary(method_upper, uri, body)
      response, body_str, attempts = perform_with_retries(method_upper, uri, body, opts, request_summary)

      record_rate_limit(response)
      @last_request_id = response['x-request-id'] || response['X-Request-Id']
      status = response.code.to_i

      if status >= 200 && status < 300
        return RawResponse.new(
          status: status,
          body: body_str,
          content_type: response['content-type'] || response['Content-Type'],
          content_disposition: response['content-disposition'] || response['Content-Disposition'],
          request_id: @last_request_id
        )
      end

      raise ApiError.from_response(response, body_str, request_summary, attempts: attempts)
    end

    private

    def execute_request(method_upper, uri, body, opts)
      cache_key = cache_key_for(method_upper, uri, opts)
      cached = cache_lookup(method_upper, cache_key, opts)
      return cached unless cached.nil?

      request_summary = build_request_summary(method_upper, uri, body)
      response, body_str, attempts = perform_with_retries(method_upper, uri, body, opts, request_summary)

      record_rate_limit(response)
      @last_request_id = response['x-request-id'] || response['X-Request-Id']

      status = response.code.to_i
      if status >= 200 && status < 300
        parsed = parse_json(body_str)
        cache_store(method_upper, cache_key, parsed, opts)
        bust_cache_for_mutation(method_upper, uri)
        return parsed
      end

      raise ApiError.from_response(response, body_str, request_summary, attempts: attempts)
    end

    def default_user_agent
      "tesote-sdk-rb/#{TesoteSdk::VERSION} (ruby/#{RUBY_VERSION})"
    end

    def build_uri(path, query)
      joined = "#{base_url}/#{version_segment}/#{path.to_s.sub(%r{\A/+}, '')}"
      uri = URI.parse(joined)
      if query && !query.empty?
        uri.query = URI.encode_www_form(stringify_query(query))
      end
      uri
    end

    def build_unversioned_uri(path, query)
      joined = "#{base_url}/#{path.to_s.sub(%r{\A/+}, '')}"
      uri = URI.parse(joined)
      if query && !query.empty?
        uri.query = URI.encode_www_form(stringify_query(query))
      end
      uri
    end

    def stringify_query(query)
      query.each_with_object([]) do |(key, value), acc|
        next if value.nil?

        if value.is_a?(Array)
          value.each { |v| acc << [key.to_s, v.to_s] }
        else
          acc << [key.to_s, value.to_s]
        end
      end
    end

    def perform_with_retries(method, uri, body, opts, request_summary)
      attempt = 0
      idempotency_key = opts[:idempotency_key]
      if MUTATING_METHODS.include?(method) && idempotency_key.nil?
        idempotency_key = SecureRandom.uuid
      end

      loop do
        attempt += 1
        begin
          response, body_str = send_once(method, uri, body, opts, idempotency_key)
        rescue Net::OpenTimeout, Net::ReadTimeout => e
          raise wrap_timeout(e, request_summary, attempt) if !retryable_network?(method) || attempt >= max_attempts

          sleep_for(backoff_delay(attempt))
          next
        rescue OpenSSL::SSL::SSLError => e
          raise TlsError.new("TLS error: #{e.message}", request_summary: request_summary, attempts: attempt, cause: e)
        rescue Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::EHOSTUNREACH,
               Errno::ENETUNREACH, SocketError, EOFError => e
          raise wrap_network(e, request_summary, attempt) if attempt >= max_attempts

          sleep_for(backoff_delay(attempt))
          next
        end

        status = response.code.to_i
        if RETRYABLE_STATUSES.include?(status) && attempt < max_attempts
          sleep_for(retry_after_for(response, attempt))
          next
        end

        return [response, body_str, attempt]
      end
    end

    def send_once(method, uri, body, opts, idempotency_key)
      request_obj = build_request(method, uri, body, opts, idempotency_key)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == 'https')
      http.open_timeout = open_timeout
      http.read_timeout = read_timeout

      log(:request, method: method, path: uri.request_uri, headers: redact_headers(request_obj))
      response = http.request(request_obj)
      body_str = response.body.to_s
      log(:response, status: response.code, request_id: response['x-request-id'])
      [response, body_str]
    end

    def build_request(method, uri, body, opts, idempotency_key)
      klass = case method
              when 'GET' then Net::HTTP::Get
              when 'POST' then Net::HTTP::Post
              when 'PUT' then Net::HTTP::Put
              when 'PATCH' then Net::HTTP::Patch
              when 'DELETE' then Net::HTTP::Delete
              else raise ArgumentError, "unsupported HTTP method: #{method}"
              end
      req = klass.new(uri.request_uri)
      req['Authorization'] = "Bearer #{api_key}"
      req['Accept'] = 'application/json'
      req['User-Agent'] = user_agent
      req['Idempotency-Key'] = idempotency_key if idempotency_key
      apply_extra_headers(req, opts[:extra_headers])
      if MUTATING_METHODS.include?(method) && !body.nil?
        req['Content-Type'] = 'application/json'
        req.body = body.is_a?(String) ? body : JSON.generate(body)
      end
      req
    end

    def apply_extra_headers(req, headers)
      return if headers.nil?

      headers.each { |k, v| req[k.to_s] = v.to_s }
    end

    def retryable_network?(method)
      # why: only retry timeouts on idempotent methods unless an idempotency key was sent
      return true if %w[GET HEAD OPTIONS PUT DELETE].include?(method)

      false
    end

    def wrap_network(error, request_summary, attempts)
      NetworkError.new("network error: #{error.class}: #{error.message}",
                       request_summary: request_summary,
                       attempts: attempts,
                       cause: error)
    end

    def wrap_timeout(error, request_summary, attempts)
      TimeoutError.new("timeout: #{error.class}: #{error.message}",
                       request_summary: request_summary,
                       attempts: attempts,
                       cause: error)
    end

    def backoff_delay(attempt)
      capped = [base_delay * (2**(attempt - 1)), max_delay].min
      jitter = @randomizer.call(capped)
      [capped + jitter, max_delay].min
    end

    def retry_after_for(response, attempt)
      header = response['retry-after'] || response['Retry-After']
      return Integer(header) if header && header.match?(/\A\d+\z/)

      backoff_delay(attempt)
    end

    def sleep_for(seconds)
      return if seconds.nil? || seconds <= 0

      @sleeper.call(seconds)
    end

    def record_rate_limit(response)
      limit = response['x-ratelimit-limit'] || response['X-RateLimit-Limit']
      remaining = response['x-ratelimit-remaining'] || response['X-RateLimit-Remaining']
      reset = response['x-ratelimit-reset'] || response['X-RateLimit-Reset']
      return if limit.nil? && remaining.nil? && reset.nil?

      @last_rate_limit = RateLimitInfo.new(
        limit: parse_int(limit),
        remaining: parse_int(remaining),
        reset: parse_int(reset)
      )
    end

    def parse_int(value)
      return nil if value.nil?

      Integer(value)
    rescue ArgumentError, TypeError
      nil
    end

    def parse_json(body_str)
      return nil if body_str.nil? || body_str.empty?

      JSON.parse(body_str)
    rescue JSON::ParserError
      body_str
    end

    def build_request_summary(method, uri, body)
      {
        method: method,
        path: uri.path,
        query: redact_query(uri.query),
        body_shape: describe_body_shape(body)
      }
    end

    def redact_query(query_string)
      return nil if query_string.nil? || query_string.empty?

      URI.decode_www_form(query_string).to_h { |(k, v)| [k, redact_secret_value(k, v)] }
    end

    def redact_secret_value(key, value)
      return '[REDACTED]' if key.to_s.match?(/key|token|secret|password/i)

      value
    end

    def redact_headers(req)
      req.each_capitalized.to_h do |key, value|
        if key.casecmp('authorization').zero?
          [key, redact_bearer(value)]
        else
          [key, value]
        end
      end
    end

    def redact_bearer(value)
      return value unless value.is_a?(String) && value.start_with?('Bearer ')

      token = value.sub(/\ABearer\s+/, '')
      last4 = token[-4..] || ''
      "Bearer ****#{last4}"
    end

    def describe_body_shape(body)
      return nil if body.nil?
      return { type: 'string', bytes: body.bytesize } if body.is_a?(String)
      return { type: 'array', items: body.size } if body.is_a?(Array)
      return { type: 'hash', keys: body.keys.map(&:to_s) } if body.is_a?(Hash)

      { type: body.class.name }
    end

    def cache_key_for(method, uri, opts)
      return nil if cache_backend.nil?
      return nil unless method == 'GET'
      return nil if opts[:cache] == false
      return nil unless opts[:cache].is_a?(Hash)

      [method, uri.path, uri.query.to_s, api_key_fingerprint].join('|')
    end

    def cache_lookup(_method, cache_key, opts)
      return nil if cache_key.nil?
      return nil if opts[:cache] == false

      cache_backend.read(cache_key)
    end

    def cache_store(method, cache_key, parsed, opts)
      return if cache_key.nil?
      return unless method == 'GET'
      return unless opts[:cache].is_a?(Hash)

      ttl = opts[:cache][:ttl] || opts[:cache]['ttl']
      return if ttl.nil?

      cache_backend.write(cache_key, parsed, ttl: ttl)
    end

    def bust_cache_for_mutation(method, _uri)
      return if cache_backend.nil?
      return unless MUTATING_METHODS.include?(method)
      return unless cache_backend.respond_to?(:clear)

      cache_backend.clear
    end

    def api_key_fingerprint
      # why: avoid cross-tenant cache bleed; do not store the raw key
      @api_key_fingerprint ||= begin
        last4 = api_key[-4..] || api_key
        "key-#{last4}"
      end
    end

    def log(event, payload)
      return if logger.nil?

      # why: a misbehaving user logger must never break the request path
      begin
        logger.call(event, payload)
      rescue NoMethodError, ArgumentError, TypeError
        nil
      end
    end
  end
end
